###Кросс-валидация

проводится в два этапа:

1. Создается генератор разбиений sklearn.model_selection.KFold, который задает набор разбиений на обучение и валидацию. Число блоков в кросс-валидации определяется параметром n_splits. Обратите внимание, что порядок следования объектов в выборке может быть неслучайным, это может привести к смещенности кросс-валидационной оценки. Чтобы устранить такой эффект, объекты выборки случайно перемешивают перед разбиением на блоки. Для перемешивания достаточно передать генератору KFold параметр shuffle=True.

2. Вычислить качество на всех разбиениях можно при помощи функции sklearn.model_selection.cross_val_score. В качестве параметра estimator передается классификатор, в качестве параметра cv — генератор разбиений с предыдущего шага. С помощью параметра scoring можно задавать меру качества, по умолчанию в задачах классификации используется доля верных ответов (accuracy). Результатом является массив, значения которого нужно усреднить.

###Задание

Нужно подобрать оптимальное значение k для алгоритма kNN. Будем использовать набор данных Wine, где требуется предсказать сорт винограда, из которого изготовлено вино, используя результаты химических анализов.

In [ ]:
import numpy as np
import io
import pandas as pd

In [ ]:
from google.colab import files
uploaded = files.upload()

Saving wine.data to wine.data


In [ ]:
names_wine = ['class', 'Alcohol', 'Malic acid', 'Ash', 
         'Alcalinity of ash', 'Magnesium', 'Total phenols',
         'Flavanoids', 'Nonflavanoid phenols', 'Proanthocyanins',
         'Color intensity', 'Hue', 'OD280/OD315 of diluted wines', 'Proline']

In [ ]:
from tables.table import Column
data = pd.read_csv(io.BytesIO(uploaded['wine.data']), names=names_wine)

In [ ]:
data.head()

class  Alcohol  Malic acid   Ash  Alcalinity of ash  Magnesium  \
0      1    14.23        1.71  2.43               15.6        127   
1      1    13.20        1.78  2.14               11.2        100   
2      1    13.16        2.36  2.67               18.6        101   
3      1    14.37        1.95  2.50               16.8        113   
4      1    13.24        2.59  2.87               21.0        118   

   Total phenols  Flavanoids  Nonflavanoid phenols  Proanthocyanins  \
0           2.80        3.06                  0.28             2.29   
1           2.65        2.76                  0.26             1.28   
2           2.80        3.24                  0.30             2.81   
3           3.85        3.49                  0.24             2.18   
4           2.80        2.69                  0.39             1.82   

   Color intensity   Hue  OD280/OD315 of diluted wines  Proline  
0             5.64  1.04                          3.92     1065  
1             4.38  1.05                          3.40     1050  
2             5.68  1.03                          3.17     1185  
3             7.80  0.86                          3.45     1480  
4             4.32  1.04                          2.93      735

In [ ]:
type(data)

pandas.core.frame.DataFrame

In [ ]:
Y_train = data[['class']]

In [ ]:
Y_train.head()

class
0      1
1      1
2      1
3      1
4      1

In [ ]:
Y_train['class'].unique()

array([1, 2, 3])

In [ ]:
X_train = data[names_wine[1:]]

In [ ]:
X_train.head()

Alcohol  Malic acid   Ash  Alcalinity of ash  Magnesium  Total phenols  \
0    14.23        1.71  2.43               15.6        127           2.80   
1    13.20        1.78  2.14               11.2        100           2.65   
2    13.16        2.36  2.67               18.6        101           2.80   
3    14.37        1.95  2.50               16.8        113           3.85   
4    13.24        2.59  2.87               21.0        118           2.80   

   Flavanoids  Nonflavanoid phenols  Proanthocyanins  Color intensity   Hue  \
0        3.06                  0.28             2.29             5.64  1.04   
1        2.76                  0.26             1.28             4.38  1.05   
2        3.24                  0.30             2.81             5.68  1.03   
3        3.49                  0.24             2.18             7.80  0.86   
4        2.69                  0.39             1.82             4.32  1.04   

   OD280/OD315 of diluted wines  Proline  
0                          3.92     1065  
1                          3.40     1050  
2                          3.17     1185  
3                          3.45     1480  
4                          2.93      735

In [ ]:
from pandas.core.common import random_state
from sklearn.model_selection import KFold, cross_val_score
from sklearn.neighbors import KNeighborsClassifier

In [ ]:
kfold = KFold(n_splits=5, shuffle=True, random_state=42)

In [ ]:
def test_accuracy(kf, X, y):
    means = list()
    for k in range(1, 51):
      knn = KNeighborsClassifier(n_neighbors=k)
      score = cross_val_score(knn, X, y, cv=kf, scoring='accuracy')
      score_mean = np.mean(score)
      means.append(score_mean)
    return pd.DataFrame(means, range(1, 51)).mean(axis=1).sort_values(ascending=False) 


In [ ]:
accuracy_wine = test_accuracy(kfold, X_train, Y_train)

/usr/local/lib/python3.7/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/usr/local/lib/python3.7/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/usr/local/lib/python3.7/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/usr/local/lib/python3.7/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for exam

In [ ]:
accuracy_wine.head()

1     0.730476
34    0.724603
35    0.724603
48    0.719048
36    0.713492
dtype: float64

k = 1 accuracy = 0.73

In [ ]:
from sklearn import preprocessing

In [ ]:
X_train_scaled = preprocessing.scale(X_train)

In [ ]:
X_train_scaled

array([[ 1.51861254, -0.5622498 ,  0.23205254, ...,  0.36217728,
         1.84791957,  1.01300893],
       [ 0.24628963, -0.49941338, -0.82799632, ...,  0.40605066,
         1.1134493 ,  0.96524152],
       [ 0.19687903,  0.02123125,  1.10933436, ...,  0.31830389,
         0.78858745,  1.39514818],
       ...,
       [ 0.33275817,  1.74474449, -0.38935541, ..., -1.61212515,
        -1.48544548,  0.28057537],
       [ 0.20923168,  0.22769377,  0.01273209, ..., -1.56825176,
        -1.40069891,  0.29649784],
       [ 1.39508604,  1.58316512,  1.36520822, ..., -1.52437837,
        -1.42894777, -0.59516041]])

In [ ]:
accuracy_wine = test_accuracy(kfold, X_train_scaled, Y_train)

/usr/local/lib/python3.7/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/usr/local/lib/python3.7/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/usr/local/lib/python3.7/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/usr/local/lib/python3.7/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for exam

In [ ]:
accuracy_wine.head()

29    0.977619
16    0.972063
15    0.972063
17    0.966508
20    0.966508
dtype: float64

После масштабирования: k = 29 accuracy = 0.978


###Задание по программированию: Выбор метрики

In [ ]:
from sklearn.neighbors import KNeighborsRegressor
import sklearn.datasets
from sklearn.preprocessing import scale
from pandas.core.common import random_state
from sklearn.model_selection import KFold, cross_val_score

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
ds = sklearn.datasets.load_boston()

/usr/local/lib/python3.7/dist-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function load_boston is deprecated; `load_boston` is deprecated in 1.0 and will be removed in 1.2.

    The Boston housing prices dataset has an ethical problem. You can refer to
    the documentation of this function for further details.

    The scikit-learn maintainers therefore strongly discourage the use of this
    dataset unless the purpose of the code is to study and educate about
    ethical issues in data science and machine learning.

    In this special case, you can fetch the dataset from the original
    source::

        import pandas as pd
        import numpy as np


        data_url = "http://lib.stat.cmu.edu/datasets/boston"
        raw_df = pd.read_csv(data_url, sep="\s+", skiprows=22, header=None)
        data = np.hstack([raw_df.values[::2, :], raw_df.values[1::2, :2]])
        target = raw_df.values[1::2, 2]

    Alternative datasets include the California housing dataset (i.e.

In [ ]:
X_train = ds.data
Y_train = ds.target

In [ ]:
X_train_scaled = sklearn.preprocessing.scale(X_train)

In [ ]:
kfold = KFold(n_splits=5, shuffle=True, random_state=42)
mink = np.linspace(1,10,num=200)
means = dict()

In [ ]:
mink

array([ 1.        ,  1.04522613,  1.09045226,  1.13567839,  1.18090452,
        1.22613065,  1.27135678,  1.31658291,  1.36180905,  1.40703518,
        1.45226131,  1.49748744,  1.54271357,  1.5879397 ,  1.63316583,
        1.67839196,  1.72361809,  1.76884422,  1.81407035,  1.85929648,
        1.90452261,  1.94974874,  1.99497487,  2.04020101,  2.08542714,
        2.13065327,  2.1758794 ,  2.22110553,  2.26633166,  2.31155779,
        2.35678392,  2.40201005,  2.44723618,  2.49246231,  2.53768844,
        2.58291457,  2.6281407 ,  2.67336683,  2.71859296,  2.7638191 ,
        2.80904523,  2.85427136,  2.89949749,  2.94472362,  2.98994975,
        3.03517588,  3.08040201,  3.12562814,  3.17085427,  3.2160804 ,
        3.26130653,  3.30653266,  3.35175879,  3.39698492,  3.44221106,
        3.48743719,  3.53266332,  3.57788945,  3.62311558,  3.66834171,
        3.71356784,  3.75879397,  3.8040201 ,  3.84924623,  3.89447236,
        3.93969849,  3.98492462,  4.03015075,  4.07537688,  4.12

In [ ]:
for p in mink:
  knn = KNeighborsRegressor(n_neighbors=5, weights='distance',p=p)
  score = cross_val_score(knn, X_train_scaled, Y_train, cv=kfold, scoring='neg_mean_squared_error')
  score_mean = np.mean(score)
  means[p] = score_mean

In [ ]:
key = max(means, key=means.get)
key

1.0

In [ ]:
means[key]

-16.030646734221644

p = 1 accuracy = -16.03

###Задание по программированию: Нормализация признаков

In [ ]:
import numpy as np
from sklearn.linear_model import Perceptron
from sklearn.preprocessing import StandardScaler

In [ ]:
X = np.array([[1, 2], [3, 4], [5, 6]])
y = np.array([0, 1, 0])
clf = Perceptron()
clf.fit(X, y)
predictions = clf.predict(X)

In [ ]:
predictions

array([0, 0, 0])

In [ ]:
scaler = StandardScaler()
X_train = np.array([[100.0, 2.0], [50.0, 4.0], [70.0, 6.0]])
X_test = np.array([[90.0, 1], [40.0, 3], [60.0, 4]])
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)


Загрузите обучающую и тестовую выборки из файлов perceptron-train.csv и perceptron-test.csv. Целевая переменная записана в первом столбце, признаки — во втором и третьем.

In [ ]:
from google.colab import files
uploaded = files.upload()

Saving perceptron-test.csv to perceptron-test (1).csv


In [ ]:
import io
import pandas as pd

In [ ]:
df_train = pd.read_csv(io.BytesIO(uploaded['perceptron-train.csv']), header=None)

In [ ]:
df_test = pd.read_csv(io.BytesIO(uploaded['perceptron-test.csv']),  header=None)

In [ ]:
df_train.head()

0         1            2
0 -1.0 -0.024626  1174.600238
1  1.0 -0.978058  1083.198803
2 -1.0  0.314272 -1472.977609
3 -1.0  0.179752   231.017267
4  1.0 -1.262544  -778.271726

In [ ]:
X_train = df_train[df_train.columns[1:3]]
y_train = df_train[df_train.columns[0]]

In [ ]:
X_test = df_test[df_test.columns[1:3]]
y_test = df_test[df_test.columns[0]]

In [ ]:
clfr = Perceptron(random_state=241)

In [ ]:
clfr.fit(X_train, y_train)

Perceptron(random_state=241)

In [ ]:
from sklearn.metrics import accuracy_score

In [ ]:
predictions = clfr.predict(X_test)

In [ ]:
accuracy = accuracy_score(y_test, predictions)

In [ ]:
accuracy

0.655

In [ ]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()

In [ ]:
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [ ]:
clfr.fit(X_train_scaled,y_train.values.ravel())

Perceptron(random_state=241)

In [ ]:
predictions_norm = clfr.predict(X_test_scaled)

In [ ]:
accuracy_norm = accuracy_score(y_test.values.ravel(), predictions_norm)

In [ ]:
accuracy_norm

0.725

In [ ]:
delta = accuracy_norm - accuracy
delta

0.06999999999999995